src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,107 @@
+        ui.greet();
+        boolean isTerminated = false;
+        while (!isTerminated) {
+            try {
+                String input = ui.readInput();
+                Command command = parser.parseCommand(input);
+                switch (command) {
+                case TODO:
+                    ToDo toDo = parser.parseToDo(input);
+                    taskList.addTask(toDo);
+                    storage.writeToFile(taskList);
+                    ui.printTaskAddedConfirmation(toDo, taskList);
+                    break;
+                case DEADLINE:
+                    Deadline deadline = parser.parseDeadline(input);
+                    taskList.addTask(deadline);
+                    storage.writeToFile(taskList);
+                    ui.printTaskAddedConfirmation(deadline, taskList);
+                    break;
+                case EVENT:
+                    Event event = parser.parseEvent(input);
+                    taskList.addTask(event);
+                    storage.writeToFile(taskList);
+                    ui.printTaskAddedConfirmation(event, taskList);
+                    break;
+                case LIST:
+                    ui.listTasks(taskList);
+                    break;
+                case FIND:
+                    String keyword = parser.parseKeyword(input);
+                    TaskList matchingTasks = taskList.findMatchingTasks(keyword);
+                    ui.printMatchingTasks(matchingTasks);
+                    break;
+                case DONE:
+                    int doneIndex = parser.parseDone(input, taskList);
+                    Task doneTask = taskList.getTask(doneIndex);
+                    taskList.markTaskAsDone(doneIndex);
+                    storage.writeToFile(taskList);
+                    ui.printTaskDoneConfirmation(doneTask);
+                    break;
+                case DELETE:
+                    int deleteIndex = parser.parseDelete(input, taskList);
+                    Task deleteTask = taskList.getTask(deleteIndex);
+                    taskList.deleteTask(deleteIndex);
+                    storage.writeToFile(taskList);
+                    ui.printTaskDeleteConfirmation(taskList, deleteTask);
+                    break;
+                case BYE:
+                    ui.printByeMessage();
+                    isTerminated = true;
+                    break;
+                default:
+                    //Do nothing
+                }
+            } catch (DukeException dukeException) {
+                ui.printDukeExceptionMessage(dukeException);
+            }
+        }
+    }
+
+    /**
+     * Initialises a Duke object after loading a saved file
+     * of Tasks if it exists, then executes main logic.
+     *
+     * @param args unused.
+     */
+    public static void main(String[] args) {
+        Ui ui = new Ui();
+        Storage storage = new Storage();
+        Parser parser = new Parser();
+        TaskList taskList = storage.readFromFile();
```

Comment:
I am not sure if this is necessary or a mistake, but following what the description of the level A-MoreOOP said, consider intitalizing these objects in the Duke constructor so the main is just treated as a main entry point instead where u can just initialize the Duke object and call the run method.